<p align="left">
  <img src="media/logo.png" width="50%"/>
</p>

[![PyPI version](https://badge.fury.io/py/mgatk.svg)](https://pypi.python.org/pypi/mgatk)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Downloads](https://pepy.tech/badge/mgatk/month)](https://pepy.tech/project/mgatk)

[Source code is made freely available](http://github.com/caleblareau/mgatk)
and a packaged install version is provided through [PyPi](https://pypi.python.org/pypi/mgatk/).
<br>

## About
This repository houses the **mgatk** package, a python-based command line interface for
processing `.bam` files with mitochondrial reads and generating high-quality heteroplasmy 
estimation from sequencing data. This package places a special emphasis on mitochondrial
genotypes generated from single-cell genomics data, primarily `mtscATAC-seq`, but is generally
applicable across other assays. 
<br>

Check out the [**mgatk** documentation](https://github.com/caleblareau/mgatk/wiki) for more 
information and user guides. 

## Contact

The **mgatk** package was developed and is maintained by [Caleb Lareau](https://profiles.stanford.edu/caleb-lareau).

<br><br>

